asm.s

```diff
@@ -62,6 +62,38 @@ PreResetTrampoline:
   # set LR to the initial value used by the ARMv7-M (0xFFFF_FFFF)
   ldr r0,=0xffffffff
   mov lr,r0
+
+  # run the pre-init code
+  bl __pre_init
+
+  # initialize .data and .bss memory
+  ldr r0,=__sbss
+  ldr r1,=__ebss
+  ldr r2,=0
+0:
+  cmp r1, r0
+  beq 1f
+  stm r0!, {r2}
+  b 0b
+1:
+
+  # copy to here
+  ldr r0,=__sdata
+  # ...up to here
+  ldr r1,=__edata
+  # copy from here
+  ldr r2,=__sidata
+2:
+  cmp r1, r0
+  beq 3f
+  # load 1 word from r2 to r3, inc r2
+  ldm r2!, {r3}
```

Comment:
Why not utilize more registers here?

Reply:
Why bother?

Reply:
Well… I haven't done any measurements, but intuition says it may be faster – as in, have less trouble saturating available memory bandwidth.

I tried to phrase this as a mild suggestion ^^

Reply:
Yeah we can certainly optimize this for more speed if we want, but the existing impl is optimized for code size.

I'm not sure how we can utilize MPU requirements here though, this code runs before the MPU can be enabled (and some chips we target don't have an MPU).

Reply:
I think simplicity is key here. At this point we're running with default clocks anyway, so it's not going to be terribly fast whatever we might do.

Reply:
The reason why MPU came up here is because it _might_ be enabled here (by e.g. a bootloader that could've been run before this), and MPU is the only thing AFAIK that would prevent code copying algorithm that goes OOB invalid.

As another point towards validity of using multiple registers, regardless of OOB reads/writes, IIRC IAR's initialization code also uses multiple registers for their thumb initialization code. Though they may have more stringent alignment requirements for their sections too.

> the existing impl is optimized for code size

Fair point. There isn't any additional cost in specifying more registers to `stm` and `ldm` instructions, but I think you'd still need some code to align the pointers a in a very specific way before the loop.

Reply:
> some code to align the pointers a in a very specific way before the loop.

The pointers can be aligned manually in the linker script, so no modification to the code is needed.